Terminal.Gui/Drawing/Aligner.cs

```diff
@@ -328,7 +328,7 @@ int spacesToGive
         return positions;
     }
 
-    private static void CheckSizeCannotBeNegative (int i, IReadOnlyList<int> sizes)
+    private static void CheckSizeCannotBeNegative (int i, ref readonly int [] sizes)
```

Comment:
These are very different things.

Aside from the other comment, the meaning of these two things are:

 - `IReadOnlyList<int>`: Any type implementing the interface, with a guarantee that, aside from unsafe memory shenanigans, the contents cannot and will not be altered and that, in the callee, the original reference can't be altered because the reference itself was passed by value, because that's how C# works.
 - `ref readonly int []`: Can only be a `System.Array<int>`, passed in by read-only reference. How's this different?
   - The parameter is an actual reference to the source array (if the compiler doesn't make a defensive copy - which it shouldn't here).
   - Normal `ref` would have allowed reassignment of the reference, but `readonly` prohibits that (`in` achieves the same in this case and is also callsite compatible, as you found out).
   - It MUST be (a reference to) an `int[]` and no other type is allowed unless it directly inherits from `System.Array<int>`, which isn't possible in human-generated code in C# - and not only because `Array<T>` isn't a "real" type you can access.
   - Typically, you'd only bother ref passing a ref type for certain interop scenarios or when you really do want the callee to be able to reassign the caller's reference to that item, which is perfectly legal and valid, but typically a very specific micro-optimization for very specific situations when in-lining will probably happen anyway.

The original was honestly probably a better type for the method argument. Interfaces are great for these sorts of things.

Reply:
Thanks.